Various fixes and improvements

ynezz opened the pull request:
* Fix possible sign conversion compiler warning
* Fix compilation without internal timer
* Make build time config option usable without patching

smuellerDD commented:
Am Mittwoch, dem 09.06.2021 um 04:47 -0700 schrieb Petr Štetiar:
> * Fix possible sign conversion compiler warning
> * Fix compilation without internal timer
> * Make build time config option usable without patching
> You can view, comment on, or merge this pull request online at:

Thanks for the post.

I hope I can get to it later this week - I will first check the other pending
items and then look at your update.

Thanks
Stephan
> 
>   https://github.com/smuellerDD/jitterentropy-library/pull/44
> 
> -- Commit Summary --
> 
>   * Make build time config option usable without patching
>   * Fix compilation without internal timer
>   * Fix possible sign conversion compiler warning
> 
> -- File Changes --
> 
>     M Makefile (35)
>     M jitterentropy-base.c (6)
>     M jitterentropy.h (38)
> 
> -- Patch Links --
> 
> https://github.com/smuellerDD/jitterentropy-library/pull/44.patch
> https://github.com/smuellerDD/jitterentropy-library/pull/44.diff
>

joshuaehill commented:
Using unsigned values for shifting and putting an `#ifdef` block around the use of the `jent_force_internal_timer` variable  is also implemented in PR #43  (and the closed PR #38 / PR #39 from which it derives).

I'm torn on using the Makefile for configuration, because it makes it somewhat more difficult to determine what the code is doing without separately reviewing the Makefile, which is not commonly distributed with the compiled library. At present, if you have have the development package installed, you can review `jitterentropy.h`, and that will tell you what features are in use. If all of this configuration information is moved to the Makefile (which is surely not going to be distributed in the case of a binary package), the the user doesn't really know how the library is going to behave.

If we are going to move toward a Makefile-based configuration, then this should probably be a bit broader (at lease including `JENT_MIN_OSR`) and there should be a programmatic interface added that reports on the various entropy-relevant parameters.

joshuaehill commented:
After applying all of my PRs, I think that such a status structure might look something like this:
```
struct rand_param
{
        unsigned int version;
        unsigned int osr;
        size_t memory_block_count; // JENT_MEMORY_BLOCKS
        size_t memory_block_size; // JENT_MEMORY_BLOCKSIZE
        size_t startup_test_values; // JENT_POWERUP_TESTLOOPCOUNT
        unsigned int apt_cutoff; // apt_cutoff
        unsigned int lag_local_cutoff; // local lag_cutoff
        unsigned int lag_global_cutoff; // global lag_cutoff
        size_t lag_window_size; //JENT_LAG_WINDOW_SIZE
        size_t lag_history_size; //JENT_LAG_HISTORY_SIZE
        unsigned int max_hash_loops; // (1U<<MIN_HASH_LOOP) + (1U<<MAX_HASH_LOOP) - 1
        unsigned int min_hash_loops; // (1U<<MIN_HASH_LOOP)
        unsigned int max_memaccess_loops; // JENT_MEMORY_ACCESSLOOPS + (1U<<MIN_ACC_LOOP_BIT) + (1U<<MAX_ACC_LOOP_BIT) - 1
        unsigned int min_memaccess_loops; // JENT_MEMORY_ACCESSLOOPS + (1U<<MIN_ACC_LOOP_BIT)
        unsigned int entropy_safety_factor; // ENTROPY_SAFETY_FACTOR
        unsigned int memaccess_enabled:1; /* Is the memaccess source configured to be used? */
        unsigned int internal_timer_enabled:1; /* Is the pthreads timer being used? */
        unsigned int shuffle_enabled:1; /* Is the pseudo random shuffle masking the underlying source? */
        unsigned int fips_enabled:1; /* Is this context in its FIPS 140 / SP 800-90B mode?? */
        unsigned int health_failure:1; /* Has this context experienced a health test failure? */
};
```

and a corresponding function would look like this:
```
JENT_PRIVATE_STATIC
void jent_status(struct rand_data *ec, struct rand_param *param)
{
        param->version = jent_version();
        param->osr = ec->osr;
        param->memory_block_count = JENT_MEMORY_BLOCKS;
        param->memory_block_size = JENT_MEMORY_BLOCKSIZE;
        param->startup_test_values = JENT_POWERUP_TESTLOOPCOUNT;
        param->apt_cutoff = ec->apt_cutoff;
        param->lag_local_cutoff = ec->lag_local_cutoff;
        param->lag_global_cutoff = ec->lag_global_cutoff;
        param->lag_window_size = JENT_LAG_WINDOW_SIZE;
        param->lag_history_size = JENT_LAG_HISTORY_SIZE;

#ifdef JENT_CONF_DISABLE_LOOP_SHUFFLE
        param->max_hash_loops = 1;
        param->min_hash_loops = 1;
        param->max_memaccess_loops = JENT_MEMORY_ACCESSLOOPS + 1;
        param->min_memaccess_loops = JENT_MEMORY_ACCESSLOOPS + 1;
#else
        param->max_hash_loops = (1U<<MIN_HASH_LOOP) + (1U<<MAX_HASH_LOOP) - 1;
        param->min_hash_loops = 1U<<MIN_HASH_LOOP;
        param->max_memaccess_loops = JENT_MEMORY_ACCESSLOOPS + (1U<<MIN_ACC_LOOP_BIT) + (1U<<MAX_ACC_LOOP_BIT) - 1;
        param->min_memaccess_loops = JENT_MEMORY_ACCESSLOOPS + (1U<<MIN_ACC_LOOP_BIT);
#endif
        param->entropy_safety_factor = ENTROPY_SAFETY_FACTOR;
        if(ec->mem == NULL) param->memaccess_enabled = 0;
        else param->memaccess_enabled = 1;

        param->internal_timer_enabled = ec->enable_notime;
#ifdef JENT_CONF_DISABLE_LOOP_SHUFFLE
        param->shuffle_enabled = 0;
#else
        param->shuffle_enabled = 1;
#endif
        param->fips_enabled = ec->fips_enabled;
        param->health_failure = ec->health_failure;
}
```

ynezz commented:
> Using unsigned values for shifting and putting an #ifdef block around the use of the jent_force_internal_timer variable is also implemented in PR

Ok, fine, wasn't aware about that. On the other hand it might be easier to just review and merge this simple PR. Anyway, I'll leave that decision on the reviewer.

>  because it makes it somewhat more difficult to determine what the code is doing without separately reviewing the Makefile, which is not commonly distributed with the compiled library

Fair enough, I've removed that bit and going to probably use some sed foo in order to modify the library before use. BTW the Makefile needs to know about the config options, in order to remove `-lpthread` linker option when `JENT_CONF_ENABLE_INTERNAL_TIMER` is disabled and pthread library not needed.
